File: quick.library/src/main/java/org/chris/quick/m/glide/progress/ProgressResponseBody.java
package org.chris.quick.m.glide.progress;


import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;
import okhttp3.MediaType;
import okhttp3.ResponseBody;
import okio.Buffer;
import okio.BufferedSource;
import okio.ForwardingSource;
import okio.Okio;
import okio.Source;

/**
 * Created by chenpengfei on 2016/11/9.
 */
public class ProgressResponseBody extends ResponseBody {

    private ResponseBody responseBody;
    private ProgressListener progressListener;
    private BufferedSource bufferedSource;

    public ProgressResponseBody(ResponseBody responseBody, ProgressListener progressListener) {
        this.responseBody = responseBody;
        this.progressListener = progressListener;
    }

    @Override
    public MediaType contentType() {
        return responseBody.contentType();
    }

    @Override
    public long contentLength() {

            return responseBody.contentLength();

    }

    @Override
    public BufferedSource source() {
        if (bufferedSource == null) {
            bufferedSource = Okio.buffer(source(responseBody.source()));
        }
        return bufferedSource;
    }

    private Source source(Source source) {
        return new ForwardingSource(source) {

            long totalBytesRead = 0;

            @Override
            public long read(Buffer sink, long byteCount) throws IOException {
                long bytesRead = super.read(sink, byteCount);
                totalBytesRead += bytesRead != -1 ? bytesRead : 0;
                if (progressListener != null) {
                    Map<String, Object> params = new HashMap<>();
                    params.put("totalBytesRead", totalBytesRead);
                    params.put("responseBodyLength", responseBody.contentLength());
                    params.put("isDone", bytesRead == -1);
                    Observable.just(params).observeOn(AndroidSchedulers.mainThread()).subscribeOn(Schedulers.io()).subscribe(new Consumer<Map<String, Object>>() {
                        @Override
                        public void accept(Map<String, Object> stringObjectMap) throws Exception {
                            long totalBytesRead = (long) stringObjectMap.get("totalBytesRead");
                            long responseBodyLength = (long) stringObjectMap.get("responseBodyLength");
                            boolean isDone = (boolean) stringObjectMap.get("isDone");
                            progressListener.progress(totalBytesRead, responseBodyLength, isDone);
                        }
                    });
                }
                return bytesRead;
            }
        };
    }
}

File: quick-component/src/main/java/org/quick/component/http/ProgressResponseBody.kt
package org.quick.component.http

import okhttp3.MediaType
import okhttp3.ResponseBody
import okio.*
import org.quick.component.QuickAsync
import org.quick.component.http.callback.OnProgressCallBack
import org.quick.component.utils.HttpUtils

/**
 * 下载进度监听-响应体
 */
class ProgressResponseBody(var builder: HttpService.Builder, var responseBody: ResponseBody, var onProgressCallBack: OnProgressCallBack) : ResponseBody() {

    var bufferedSource: BufferedSource

    init {
        bufferedSource = Okio.buffer(source(responseBody.source()))
    }

    override fun contentLength(): Long {
        return responseBody.contentLength() + builder.downloadStartIndex
    }

    override fun contentType(): MediaType? = responseBody.contentType()

    override fun source(): BufferedSource {
        return bufferedSource
    }

    private fun source(source: Source): Source {
        return object : ForwardingSource(source) {
            var totalBytesRead = builder.downloadStartIndex
            override fun read(sink: Buffer, byteCount: Long): Long {
                val bytesRead = super.read(sink, byteCount)
                totalBytesRead += if (bytesRead != -1L) bytesRead else 0
                QuickAsync.runOnUiThread {
                    onProgressCallBack.onLoading(HttpUtils.getFileName(builder.url), totalBytesRead, contentLength(), bytesRead == -1L)
                }
                return bytesRead
            }
        }
    }

}